assert.h

```diff
@@ -21,19 +24,31 @@
 
 namespace snowhouse
 {
+  auto WithMessage = [](const std::string& msg) { return [&]() { return msg; }; };
+
```

Comment:
I didn't realize how often we had to specify the empty message supplier. 

Let's add

```cpp
  namespace detail {
    auto EmptyMessage = []() { return std::string(); };
  }
```
and use that. That's more readable.

Reply:
Are you sure this won't violate the One Definition Rule? `EmptyMessage`  will become an object, and it will be defined in every translation unit which would have `assert.h` included, leading to multiple definitions and linker errors. The same applies to `WithMessage`\`ExtraMessage` defined as a named symbol in line 27. I might be wrong, and my VS2019 _seems_ to compile and link correctly when I add another translation unit which includes Igloo, but I am not 100% it's valid code.

Am I wrong?

Reply:
I made both `EmptyMessage` and `ExtraMessage` inline functions to avoid violation of ODR and problems with linking. They _could_ probably be made `inline` "named lambdas", but my VS2019 seems to not like `inline` variables and fails to compile them. I hesitate to make them "regular, named lambdas" because I _think_ they might violate ODR (I am not sure though, my C++ got really rusty :( ), but if you are sure they won't, I can do it.

Reply:
I don't know C++ much. It might be a violation. https://stackoverflow.com/a/62150779
I thought `assert.h` having a header guard helps, but not sure.

Reply:
> I think `inline` variables should be good.

I wanted to use them, but my VS2019 complained. Either it uses C++14 compliant compiler, or this C++17 feature is not supported, or I did something wrong.

> What do you mean by "regular, named lambdas"?

I mean variables declared as `auto name = [](args) { lanbda body; };` I am not sure how it's technically called, because _lambda_ is used just for initialization, and `name` denotes... I do not know what exactly ;) but since the thing is named, it's not a lambda anymore, right?

> Usual functor classes can use inline member functions, so I don't think they'll have ODR issues.

I used functors at the beginning, but then kazk asked to change them to variables initialized with lambdas, and it was just some tme later when I realized that such variables defined in a header can (can they really?) violate ODR.

> Are lvalue instances of `WithMessage`/`ExtraMessage` supposed to exist? I haven't looked at the code thoroughly, but with using strings by reference it looks like there will be some issues if someone creates an lvalue WithMessage from an rvalue string.

Oh I was going to point this one out too and forgot. Definitely it's not advised to store suppliers for later (re-)use, but you are right that his does not work well:

```cpp
    It(Test_Supplier_Variable) {

        auto supplier = ExtraMessage("Test failed for ya!");
        Assert::That(42, Equals(13), supplier);
        Assert::That(42, Equals(7), supplier);
    }
```

The output is not as expected:

```text
<IT::>Test_Supplier_Variable

<FAILED::>Expected: equal to 13<:LF:>Actual: 42<:LF:>

<COMPLETEDIN::>
```

Changing `ExtraMessage` to capture by value fixes the issue, as expected.

Thanks for your review, and let me know if I got you right.

Reply:
> I wanted to use them, but my VS2019 complained. Either it uses C++14 compliant compiler, or this C++17 feature is not supported, or I did something wrong.

This is the first time I've seen inline variables, but it looks like this is exactly the case they exist for. What about `static`?

> but since the thing is named, it's not a lambda anymore, right?

I guess a lambda is always a lambda. Its _class_ name is always unique and unknown.

> such variables defined in a header can (can they really?) violate ODR.

I think they can. If you repeat it in another translation unit, it's the second definition, otherwise there's no declaration. Normally, a definition should be in a separate translation unit, but it's a header-only library.

Reply:
Anyway, I think that `EmptyMessage` and `ExtraMessage` are perfectly fine as they are: inline, free functions. I'd say there's no need, and no benefit, in making them lambda-initialized variables, be it marked as inline or static or whatever, just for sake of turning functions into lambda expressions. Is there anything in particular better in

```cpp
    inline auto EmptyMessage = []() { return std::string(); };
```

vs

```cpp
    inline std::string EmptyMessage()
    { 
      return std::string();
    }
```

I personally think there's not.

Reply:
To be clear, I don't care if these are lambdas or functions, and I don't think there's a significant difference we'd care.

Originally suggested changing because I thought structs were unnecessary, especially since those were brought into scope. The lambda version compiled and worked as expected, so I thought that was good (spoiled by Rust).

Tested the latest version against all published C++ kata and had the same result (no error or new warning).

Reply:
> Tested the latest version against all published C++ kata and had the same result (no error or new warning).

Great! When (and if) we get it deployed, I could test it on some kata from C++ list when fixing warnings.

Reply:
I would create some docs for cpp too similar to what we have for C or JS: some kind of reference, and authoring tutorial.